docs/book/v3/features/helpers/template-variable-container.md

```diff
@@ -0,0 +1,149 @@
+section of the website, currently matched route, etc. Unfortunately, because the
+method changes the internal state of the renderer, this can cause problems in an
+async environment, such as [Swoole](https://docs.zendframework.com/zend-expressive-swoole), 
+where those changes will persist for parallel and subsequent requests.
+
+To provide a stateless alternative, you can create a `Zend\Expressive\Helper\Template\TemplateVariableContainer`
+and persist it as a request attribute. This allows you to set template variables
+that are pipeline-specific, and later extract and merge them with
+handler-specific values when rendering.
+
+To facilitate this further, we provide `Zend\Expressive\Helper\Template\TemplateVariableContainerMiddleware`,
+which will populate the attribute for you if it has not yet been.
+
+The container is **immutable**, and any changes will result in a new instance.
+As such, any middleware that is providing additional values or removing values
+**must** call `$request->withAttribute()` to replace the instance, per the
+examples below.
+
+As an example, consider the following pipeline:
+
+```php
+use Psr\Container\ContainerInterface;
+use Zend\Expressive\Application;
+use Zend\Expressive\Handler\NotFoundHandler;
+use Zend\Expressive\Helper\ServerUrlMiddleware;
+use Zend\Expressive\Helper\Template\TemplateVariableContainerMiddleware;
+use Zend\Expressive\Helper\UrlHelperMiddleware;
+use Zend\Expressive\MiddlewareFactory;
+use Zend\Expressive\Router\Middleware\DispatchMiddleware;
+use Zend\Expressive\Router\Middleware\ImplicitHeadMiddleware;
+use Zend\Expressive\Router\Middleware\ImplicitOptionsMiddleware;
+use Zend\Expressive\Router\Middleware\MethodNotAllowedMiddleware;
+use Zend\Expressive\Router\Middleware\RouteMiddleware;
+use Zend\Stratigility\Middleware\ErrorHandler;
+
+use function Zend\Stratigility\path;
+
+return function (Application $app, MiddlewareFactory $factory, ContainerInterface $container) : void {
+    $app->pipe(ErrorHandler::class);
+    $app->pipe(ServerUrlMiddleware::class);
+
+    $app->pipe(path(
+        '/api/doc',
+        $factory->lazy(TemplateVariableContainerMiddleware::class)
+    ));
+
+    $app->pipe(RouteMiddleware::class);
+
+    $app->pipe(ImplicitHeadMiddleware::class);
+    $app->pipe(ImplicitOptionsMiddleware::class);
+    $app->pipe(MethodNotAllowedMiddleware::class);
+    $app->pipe(UrlHelperMiddleware::class);
+
+    $app->pipe(DispatchMiddleware::class);
+
+    $app->pipe(NotFoundHandler::class);
+};
+```
+
+Any middleware or handler that responds to a path beginning with `/api/doc` will
+now have a `Zend\Expressive\Helper\Template\TemplateVariableContainer` attribute
+that contains an instance of that class.
+
+Within middleware that responds on that path, you can then do the following:
+
+```php
+use Zend\Expressive\Helper\Template\TemplateVariableContainer;
+use Zend\Expressive\Router\RouteResult;
+
+$container = $request->getAttribute(
```

Comment:
The entry sentence for this code example says "Within middleware", but the typical structure of a middleware is missing here. The example is too reduced and difficult for beginners to read.